### PR TITLE
Shoot care refactoring

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area quality
/kind cleanup
/priority normal

**What this PR does / why we need it**:
This PR submits a first batch of refactoring for shoot care operations:
- Extract health checks from `botanist` -> `/pkg/operation/care`
- Extract garbage collection from `botanist` -> `/pkg/operation/care`
- `Operation` is fed with shoot from `Cluster` object
- Adds and improves unit tests

**Special notes for your reviewer**:
Further refactoring is planned. Submitting these changes already should help to simplify reviews.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/client/kubernetes/fake/clientset.go [hunk 1]
```diff
@@ -149,7 +152,10 @@ func (c *ClientSet) ForwardPodPort(string, string, int, int) (chan struct{}, err
 	return nil, nil
 }
 
+// CheckCheckForwardPodPort is a type alias for a function checking port forwarding for pods.
+type CheckCheckForwardPodPort func(string, string, int, int) error
```
Comment (rfranzke) [inline pkg/client/kubernetes/fake/clientset.go]:
```suggestion
type CheckForwardPodPortFn func(string, string, int, int) error
```

**Comment by danielfoehrKn** [comment]
Looks good from my POV and works during tests.

Refactoring the care operations away from the botanist is a general motivation (botanist heavy weight & cloud-provider specific), or is there a specific reason that I am unaware of?

**Comment by timuthy** [comment]
> Looks good from my POV and works during tests.

Thanks for the review @danielfoehrKn.

> Refactoring the care operations away from the botanist is a general motivation (botanist heavy weight & cloud-provider specific), or is there a specific reason that I am unaware of?

The main points are:
- Botanist is shoot reconcile/deletion specific (IMO). We wanted to get rid of it completely after we've externalized all cloud specifics, IIRC (/cc @rfranzke)
- Restricted testability due to its size (as you mentioned)

**Comment by timuthy** [comment]
Update - PR has been rebased after merge conflicts.

File: pkg/operation/care/garbage_collection.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		shoot:                  op.Shoot,
+		initializeShootClients: shootClientInit,
+		seedClient:             op.K8sSeedClient.Client(),
+		logger:                 op.Logger,
+	}
+}
+
+// GarbageCollection cleans the Seed and the Shoot cluster from no longer required
+// objects. It receives a botanist object <botanist> which stores the Shoot object.
+func (g *GarbageCollection) GarbageCollection(ctx context.Context) {
```
Comment (rfranzke) [inline pkg/operation/care/garbage_collection.go]:
```suggestion
func (g *GarbageCollection) Collect(ctx context.Context) {
```

File: pkg/operation/care/garbage_collection.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+func (g *GarbageCollection) performGarbageCollectionSeed(ctx context.Context) error {
+	podList := &corev1.PodList{}
+	if err := g.seedClient.List(ctx, podList, client.InNamespace(g.shoot.SeedNamespace)); err != nil {
+		return err
+	}
+
+	if err := g.deleteStalePods(ctx, g.seedClient, podList); err != nil {
+		return err
+	}
+	return nil
```
Comment (rfranzke) [inline pkg/operation/care/garbage_collection.go]:
```suggestion
	return g.deleteStalePods(ctx, g.seedClient, podList)
```

File: pkg/operation/care/health.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+		shoot:                  op.Shoot,
+		seedClient:             op.K8sSeedClient,
+		initializeShootClients: shootClientInit,
+		shootClient:            op.K8sShootClient,
+		logger:                 op.Logger,
+	}
+}
+
+// HealthChecks conducts the health checks on all the given conditions.
+func (h *Health) HealthChecks(
```
Comment (rfranzke) [inline pkg/operation/care/health.go]:
```suggestion
func (h *Health) Check(
```

**Comment by rfranzke** [comment]
> Botanist is shoot reconcile/deletion specific (IMO). We wanted to get rid of it completely after we've externalized all cloud specifics, IIRC (/cc @rfranzke)

Hm, what do you mean by it? Which cloud specifics were not yet externalized?

**Comment by timuthy** [comment]
> > Botanist is shoot reconcile/deletion specific (IMO). We wanted to get rid of it completely after we've externalized all cloud specifics, IIRC (/cc @rfranzke)
> 
> Hm, what do you mean by it? Which cloud specifics were not yet externalized?

I was just saying that we had a discussion longer time ago when we were actively working on extensibility and spoke about getting rid of the "botanist" in general once we finished the extensibility topic. Anyways, I think it makes sense to extract the health checks because the `Botanist` has predominately reconcile/delete functions which are not really related to health checks.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
> > > Botanist is shoot reconcile/deletion specific (IMO). We wanted to get rid of it completely after we've externalized all cloud specifics, IIRC (/cc @rfranzke)
> > 
> > 
> > Hm, what do you mean by it? Which cloud specifics were not yet externalized?
> 
> I was just saying that we had a discussion longer time ago when we were actively working on extensibility and spoke about getting rid of the "botanist" in general once we finished the extensibility topic. Anyways, I think it makes sense to extract the health checks because the `Botanist` has predominately reconcile/delete functions which are not really related to health checks.

Hm ok, I don't remember the details, but I guess the 'Botanist' as it is today is fine. It carries some information required for shoot reconciliation/deletion. In the past, we also had the 'CloudBotanist' or the 'HybridBotanist', and both we could eliminate with the extensibility APIs.
Still, we could think about renaming 'botanist' to something else in the future :)

**Pull request merged**